File: Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if(l1==nullptr){
            return l2;
        }
        if(l2==nullptr){
            return l1;
        }
        int sum=0,carry=0;
        ListNode*x=l1;
        ListNode*y=l2;
         ListNode*z=new ListNode(-1);
         ListNode*z2=nullptr;
        while(x!=nullptr&&y!=nullptr){
            sum=x->val+y->val+carry;
            carry=sum/10;
            x=x->next;
            y=y->next;
            z->next=new ListNode(sum%10);
            if(z2==nullptr){
                z2=z->next;
            }
            z=z->next;
        }
        if(x!=nullptr){
            while(x!=nullptr){
                sum=x->val+carry;
            carry=sum/10;
            x=x->next;
            z->next=new ListNode(sum%10);
                z=z->next;
            }
        }
         if(y!=nullptr){
            while(y!=nullptr){
                sum=y->val+carry;
            carry=sum/10;
            y=y->next;
             z->next=new ListNode(sum%10);
                z=z->next;
            }
        }
        if(carry!=0){
             z->next=new ListNode(carry);
        }
        return z2;
    }
};

File: Solutions in Java/Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    // example in leetcode book
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    	ListNode dummyHead = new ListNode(0);
    	ListNode p = l1, q= l2, curr = dummyHead;
    	int carry = 0;
    	while (p != null || q!= null) {
    		int x = (p != null) ? p.val : 0;
    		int y = (q != null) ? q.val : 0;
    		int digit = carry + x + y;
    		carry = digit / 10;
    		curr.next = new ListNode(digit % 10);
    		curr = curr.next;
    		if (p != null) p = p.next;
    		if (q != null) q = q.next;
    	}
    	if (carry > 0) {
    		curr.next = new ListNode(carry);
    	}
    	return dummyHead.next;
    }
}
